### ISSUE TITLE
mailroom-ml: contract head carries a zero-train/zero-val `other` class (26-class head, 25 weights)

### ISSUE THREAD

**Issue by Exios66** [LLM-Mailroom-Services/mailroom-issues#116]
**Spawned from #112 (M9a) — data/taxonomy QA (M9a-U2).** Domain: `Mailroom ML`.

## Problem
The contract head is declared a **26-class** head (`labels.json` / `vocabularies.json`, `CONTRACT_SUBTYPE_KEYS + ("other",)`) but the pinned corpus observes **zero** `other` rows in train, validation, **and** test. Two consequences:

1. `artifacts/pytorch/model/labels.json` contract `weights` has only **25 entries** — the zero-row `other` class gets the default weight, and `macro_f1` over 26 classes includes a class with no support anywhere.
2. `observed_label_surfaces` (`src/mailroom_ml/labels.py:233`) grants contract `other` "by fiat" (preserved as the CUAD fallback token at inference) while correspondence `other` and corporate_record `certificate_of_formation` are dropped as zero-row enums. The asymmetry is intentional per the tracker, but it is undocumented in the artifact and surprises every macro-F1 reader.

## Evidence
- `labels.py:233` — `surfaces[cls] = CONTRACT_SUBTYPE_KEYS + ("other",)`.
- `labels.py:106` — `SUBCLASS_BY_CLASS["contract"] = CONTRACT_SUBTYPE_KEYS + ("other",)`.
- U2 audit: contract `other` has 0 rows in train/val/test; `weights` map has 25 keys vs the 26-`labels` list.
- U2: 5 contract val classes have **zero** support, 6 have **n==1**, 19/21 observed have n<5 → a 26-class macro-F1 is not statistically interpretable on this split.

## Decision needed (human/tracker)
Pick one and document it in the artifact `note`:
- **(a)** keep `other` as an inference-only fallback token (like doc_type `unknown`) and exclude it from train-time macro-F1; or
- **(b)** drop `other` from the training head entirely and handle it purely via the `SUBCLASS_PROJECTIONS` / unmapped-route path.

Either way, the macro-F1 aggregation should not silently average a zero-support class, and per-head metrics should always ship with per-class support (the eval harness now emits `per_head.<head>.support` — commit `1718358`).

Relates to #112, #66, #67, #75, #85.

**Comment by Exios66** [comment]
### Verification
- Verdict: **NEEDS_SCOPING**
- Evidence:
  - `src/mailroom_ml/labels.py` — contract head still `CONTRACT_SUBTYPE_KEYS + ("other",)`; zero train/val/test rows for `other` remain the design tension described in the issue.
  - `tests/test_labels.py::test_contract_other_and_doc_type_unknown_are_inference_only` — `other` ∈ `inference_only`, ∉ `weights`.
  - `tests/test_train.py::test_inference_only_class_zero_support_excluded_and_default_weight` — observed macro-F1 excludes zero-support class; documents default CE weight 1.0 trap.
  - Observed on `mailroom-ml` `main` @ `4050ee4`.

### Fix (if applicable)
- PR: **none** — requires human/tracker choice between issue options **(a)** inference-only fallback vs **(b)** drop from training head + route via projections.
- What changed: N/A pending product decision.
- Contract/test proof: N/A (policy gate, not a code defect with a single safe minimal fix).

### Close policy
- Issue stays **OPEN** until operator picks (a) or (b) and a follow-up PR lands (if any).

**Comment by Exios66** [comment]
## Decision locked (Jack / CoS) — 2026-09-24

**Choice (hybrid of tracker options a/b):**

> Keep `other` in the contract for OOV semantics and mandatory review, but remove it from the trainable decision space (**loss + argmax**); **only routing/projection** may assign `other`.

### Implications for implementers
1. `other` stays in the contract label surface (OOV / CUAD fallback / mandatory-review semantics).
2. Trainable head excludes `other` from supervised loss and from argmax over model logits.
3. Assignment of `other` is reserved for routing / `SUBCLASS_PROJECTIONS` / unmapped-route paths.
4. macro-F1 and artifact notes must not silently treat `other` as a zero-support trainable class.

Implementation: Composer 2.5 agent in flight — will open a draft PR on `mailroom-ml` and comment here with the PR link. **Do not close this issue until that PR merges with final proof.**

Agent: https://cursor.com/agents/bc-f8977de4-7a01-5826-a340-dbb91a4b66cd

**Comment by Exios66** [comment]
Decision + implementation tracking mirrored in mailroom-ml: https://github.com/LLM-Mailroom-Services/mailroom-ml/issues/15 (options a/b, trainer weights, enrichment rejects). Epic: #85 for any head-vocab change.

**Comment by Exios66** [comment]
**Decided direction (implementation by parallel agent):** Retain `other` in the contract surface for OOV / mandatory-review semantics; **exclude `other` from the trainable decision space** (subclass CE + argmax). Assignment to `other` only via routing/projection layers.

mailroom-ml tracker: https://github.com/LLM-Mailroom-Services/mailroom-ml/issues/15

**Comment by Exios66** [comment]
## Jack decision (implemented)

> Keep `other` in the contract for OOV semantics and mandatory review, but remove it from the trainable decision space (loss + argmax); only routing/projection may assign `other`.

## PR

https://github.com/LLM-Mailroom-Services/mailroom-ml/pull/31 (draft)

## Files changed

- `src/mailroom_ml/labels.py` — `routing_only`, `trainable_labels` / trainable id maps; contract artifact `note`
- `src/mailroom_ml/inference.py` — trainable-slice argmax; `normalize_label_maps` on bundle load
- `training/train_modernbert.py` — 25-wide contract head; CE `ignore_index` for routing-only targets
- `deploy/onnx_export.py` — export head sizes from `trainable_labels`
- `tests/test_labels.py`, `tests/test_train.py` — #116 pins

## Tests

```bash
python3 -m pytest -m "not fullcorpus" -q
```

**Result:** 192 passed, 35 skipped (2026-09-24 UTC cloud agent run)

## Behavior summary

| Surface | `other` treatment |
|---|---|
| `labels` / routing contract | retained (26 keys) |
| Trainable head / loss / argmax | excluded (`trainable_labels` = 25 CUAD families) |
| Routing / `project_subclass` | may still assign `other` |
| macro-F1 | zero-support `other` not in trainable denominator; `support` stays honest in eval |

Not closing this issue — CoS after merge + proof.

**Comment by Exios66** [comment]
## Final proof — closing

**PR:** https://github.com/LLM-Mailroom-Services/mailroom-ml/pull/31 (merged)
**Live merge SHA on `mailroom-ml` `main`:** `e3dc3294635c03e753712f8ad8f57a2b2ca98728`
**Verified:** `main` tip == merge SHA (2026-09-24 ~4:30 PM CT)

### Decision implemented
Keep contract `other` in the routing/label contract; exclude from trainable decision space (loss + argmax). Only routing / `SUBCLASS_PROJECTIONS` may assign `other`.

### Landed surfaces
- `labels.py` — `routing_only`, `trainable_labels` / trainable id maps; contract note
- `train_modernbert.py` — 25-wide contract head; CE `ignore_index` for routing-only
- `inference.py` — trainable-slice argmax; legacy 26-logit load path
- `deploy/onnx_export.py` — head sizes from `trainable_labels`
- Tests pin 26 labels / 25 trainable (`test_labels.py`, `test_train.py`)

Closing as completed with live merge SHA proof.
